tensorpipe/channel/cuda_ipc/channel.cc

```diff
@@ -0,0 +1,615 @@
+#include <tensorpipe/channel/helpers.h>
+#include <tensorpipe/common/callback.h>
+#include <tensorpipe/common/defs.h>
+#include <tensorpipe/common/error.h>
+#include <tensorpipe/common/error_macros.h>
+#include <tensorpipe/common/optional.h>
+#include <tensorpipe/common/queue.h>
+#include <tensorpipe/common/system.h>
+#include <tensorpipe/proto/channel/cuda_ipc.pb.h>
+
+#define TP_CUDA_CHECK(a)                                                      \
+  TP_THROW_ASSERT_IF(cudaSuccess != (a))                                      \
+      << __TP_EXPAND_OPD(a) << " " << cudaGetErrorName(cudaPeekAtLastError()) \
+      << " (" << cudaGetErrorString(cudaPeekAtLastError()) << ")"
+
+namespace tensorpipe {
+namespace channel {
+namespace cuda_ipc {
+
+namespace {
+
+class CudaEvent {
+ public:
+  explicit CudaEvent(int device) : device_(device) {
+    TP_CUDA_CHECK(cudaSetDevice(device_));
+    TP_CUDA_CHECK(cudaEventCreateWithFlags(
+        &ev_, cudaEventDisableTiming | cudaEventInterprocess));
+  }
+
+  explicit CudaEvent(cudaIpcEventHandle_t handle) {
+    TP_CUDA_CHECK(cudaIpcOpenEventHandle(&ev_, handle));
+  }
+
+  void record(cudaStream_t stream) {
+    TP_CUDA_CHECK(cudaEventRecord(ev_, stream));
+  }
+
+  void wait(cudaStream_t stream) {
+    TP_CUDA_CHECK(cudaStreamWaitEvent(stream, ev_, 0));
+  }
+
+  std::string serializedHandle() {
+    cudaIpcEventHandle_t handle;
+    TP_CUDA_CHECK(cudaIpcGetEventHandle(&handle, ev_));
+
+    return std::string(reinterpret_cast<const char*>(&handle), sizeof(handle));
+  }
+
+  ~CudaEvent() {
+    TP_CUDA_CHECK(cudaSetDevice(device_));
+    TP_CUDA_CHECK(cudaEventDestroy(ev_));
+  }
+
+ private:
+  int device_{0};
+  cudaEvent_t ev_;
+};
+
+int cudaDeviceForPointer(const void* ptr) {
+  cudaPointerAttributes attrs;
+  TP_CUDA_CHECK(cudaPointerGetAttributes(&attrs, ptr));
+  return attrs.device;
+}
+
+class SendOperation {
+ public:
+  uint64_t sequenceNumber;
+  Channel::TSendCallback callback;
+
+  SendOperation(
+      uint64_t sequenceNumber,
+      Channel::TSendCallback callback,
+      const void* ptr,
+      size_t length,
+      cudaStream_t stream)
+      : sequenceNumber(sequenceNumber),
+        callback(std::move(callback)),
+        ptr_(ptr),
+        length_(length),
+        stream_(stream) {
+    if (length > 0) {
+      startEv_ = std::make_unique<CudaEvent>(cudaDeviceForPointer(ptr));
+    }
+  }
+
+  proto::Descriptor descriptor() {
+    proto::Descriptor pbDescriptor;
+
+    if (length_ > 0) {
+      pbDescriptor.set_start_ev_handle(startEv_->serializedHandle());
+
+      cudaIpcMemHandle_t handle;
+      TP_CUDA_CHECK(cudaIpcGetMemHandle(&handle, const_cast<void*>(ptr_)));
+      pbDescriptor.mutable_handle()->append(
+          reinterpret_cast<const char*>(&handle), sizeof(handle));
+    }
+
+    return pbDescriptor;
+  }
+
+  void process(const cudaIpcEventHandle_t& stopEvHandle) {
+    if (length_ == 0) {
+      return;
+    }
+
+    startEv_->record(stream_);
```

Comment:
I'm very confused by this. SendOperation::process is being called when the sender receives the reply from the receiver. Which means that we record the start event *after* the receiver has already launched the memcpy and thus after the receiver has recorded the start event. Doesn't this mean that the receiver has recorded an empty event?

Reply:
I'll move the `record()` earlier.